apps/passport-client/pages/index.tsx

```diff
@@ -462,9 +434,83 @@ async function loadInitialState(): Promise<AppState> {
 
 registerServiceWorker();
 
-const root = createRoot(document.querySelector("#root"));
-root.render(
-  <RollbarProvider>
-    <App />
-  </RollbarProvider>
-);
+interface AppStateProviderProps {
+  children: React.ReactNode;
+  initialState: AppState;
+}
+
+const stateEmitter: StateEmitter = new Emitter();
+
+const AppStateProvider: React.FC<AppStateProviderProps> = ({
+  children,
+  initialState
+}) => {
+  const [state, setState] = useState<AppState>(initialState);
+  const [lastDiff, setLastDiff] = useState<Partial<AppState>>({});
+
+  const update = useCallback(
+    (diff: Partial<AppState>): void => {
+      setState(Object.assign(state, diff));
+
+      // In a React class component, the `setState` method has a second
+      // parameter, which is a callback function that React will invoke when
+      // the state change has taken effect. `useState` does not offer the same
+      // functionality, and the recommended approach is to use a `useEffect`
+      // hook with the relevant piece of state as a dependency. The effect hook
+      // will then be invoked whenever the state changes.
+      //
+      // However, we specifically want to observe changes to the `state`
+      // variable, and the use of `Object.assign` above ensures that, from
+      // React's perspective, the object doesn't change, at least not in a way
+      // that would trigger a re-render or an effect hook to run. This is
+      // because we do not change the object's identity, only its content.
+      //
+      // So, we need to set up some other piece of state that changes whenever
+      // the state object does. Here, we track the receipt of diffs in the
+      // update method, and in the below `useEffect` hook we trigger the hook
+      // to fire whenever a new diff is received. This allows the hook to fire
+      // on state changes even though it can't track a change to the state
+      // object directly. It will then emit an event, which is what the rest of
+      // the app uses to work around the fact that it also can't track changes
+      // to the state object.
+      setLastDiff(diff);
```

Comment:
could we instead call `stateEmitter.emit(state);` here rather than in the effect below (and eliminating the need to store this `lastDiff` value?

Reply:
Possibly. The difference is that with the old class-based `setState`, React would trigger the callback after the state change is committed. If we emit the event after we've called `setState` but before we know it has committed (which is what we're using the effect hook to detect), it might be the case that we see inconsistent state.

For instance, we use the events emitted here to tell the `usePCDCollection` selector to re-wrap the PCDCollection object. If the event fires before the relevant state change has been committed, then it will incorrect wrap the old state. My understanding is that this is why the recommended alternative to the `setState` callback is to use an effect hook. There's a discussion about it [here](https://github.com/reactjs/rfcs/issues/98) from when hooks were first introduced.